File: FixtureGenerators/PooledFinals.py
from FixtureGenerators.FixturesGenerator import FixturesGenerator
from database.models import Tournaments, Games
from structure import manage_game


# [sf1, sf2], [3v3, 4v4, 5v5]
# [sf1, 3v3, sf2, 4v4, 5v5]

class PooledFinals(FixturesGenerator):
    def __init__(self, tournament_id):
        super().__init__(tournament_id, fill_officials=True, editable=False, fill_courts=True)
    
    def _end_of_round(self, tournament_id):
        pool_one, pool_two = [[j[0] for j in i] for i in Tournaments.query.filter(Tournaments.id == tournament_id).first().ladder()]
        finals_games = Games.query.filter(Games.tournament_id == tournament_id, Games.is_final == True).all()
        last_game = Games.query.filter(Games.tournament_id == tournament_id).order_by(Games.round.desc()).first()

        if len(finals_games) > 2:
            self.end_tournament()
        elif finals_games:
            manage_game.create_game(tournament_id, finals_games[0].losing_team_id, finals_games[1].losing_team_id,
                                    is_final=True,
                                    round_number=finals_games[-1].round + 1)
            manage_game.create_game(tournament_id, finals_games[0].winning_team_id, finals_games[1].winning_team_id,
                                    is_final=True,
                                    round_number=finals_games[-1].round + 1)
        else:
            manage_game.create_game(tournament_id, pool_one[0].id, pool_two[1].id, is_final=True,
                                    round_number=last_game.round + 1)
            manage_game.create_game(tournament_id, pool_two[0].id, pool_one[1].id, is_final=True,
                                    round_number=last_game.round + 1)

File: HandballBackend/FixtureGenerator/PooledFinals.cs
using HandballBackend.EndpointHelpers;
using HandballBackend.EndpointHelpers.GameManagement;
using Microsoft.EntityFrameworkCore;

namespace HandballBackend.FixtureGenerator;

public class PooledFinals : AbstractFixtureGenerator {
    private readonly int _tournamentId;


    public PooledFinals(int tournamentId) : base(tournamentId, true, true) {
        _tournamentId = tournamentId;
    }

    public override async Task<bool> EndOfRound() {
        var db = new HandballContext();
        var tournament = (await db.Tournaments.FindAsync(_tournamentId))!;

        var finalsGames = await db.Games.Where(g => g.TournamentId == _tournamentId && g.IsFinal).OrderBy(g => g.Id)
            .ToListAsync();

        if (finalsGames.Count > 2) {
            // each round is 2 games, so > 2 means we've had both rounds
            EndTournament();
            return true;
        }

        var tasks = new List<Task>();
        if (finalsGames.Count != 0) {
            tasks.Add(GameManager.CreateGame(_tournamentId, finalsGames[0].LosingTeamId, finalsGames[1].LosingTeamId,
                isFinal: true, round: finalsGames[0].Round + 1));
            tasks.Add(GameManager.CreateGame(_tournamentId, finalsGames[0].WinningTeamId!.Value,
                finalsGames[1].WinningTeamId!.Value, isFinal: true, round: finalsGames[0].Round + 1));
        } else {
            var (_, poolOne, poolTwo) = await LadderHelper.GetTournamentLadder(db, tournament);
            var lastGame = await db.Games.Where(g => g.TournamentId == _tournamentId).OrderByDescending(g => g.Id)
                .FirstAsync();
            tasks.Add(GameManager.CreateGame(_tournamentId, poolOne![0].Id, poolTwo![1].Id, isFinal: true,
                round: lastGame.Round + 1));
            tasks.Add(GameManager.CreateGame(_tournamentId, poolTwo[0].Id, poolOne[1].Id, isFinal: true,
                round: lastGame.Round + 1));
        }

        await Task.WhenAll(tasks);
        return await base.EndOfRound();
    }
}